sklearn/cluster/_k_means_lloyd.pyx

```diff
@@ -0,0 +1,398 @@
+        if update_centers:
+            with gil:
+                for j in range(n_clusters):
+                    weight_in_clusters[j] += weight_in_clusters_chunk[j]
+                    for k in range(n_features):
+                        centers_new[j, k] += centers_new_chunk[j * n_features + k]
+
+        free(centers_new_chunk)
+        free(weight_in_clusters_chunk)
+        free(pairwise_distances_chunk)
+
+    if update_centers:
+        _relocate_empty_clusters_dense(X, sample_weight, centers_old,
+                                    centers_new, weight_in_clusters, labels)
+
+        _average_centers(centers_new, weight_in_clusters)
+        _center_shift(centers_old, centers_new, center_shift)
+
+
+cdef void _update_chunk_dense(floating *X,
+                              # expecting C alinged 2D array. XXX: Can be
+                              # replaced by const memoryview when cython min
+                              # version is >= 0.3
+                              floating[::1] sample_weight,
+                              floating[::1] x_squared_norms,
+                              floating[:, ::1] centers_old,
+                              floating[::1] centers_squared_norms,
+                              int[::1] labels,
+                              floating *centers_new,
+                              floating *weight_in_clusters,
+                              floating *pairwise_distances,
+                              bint update_centers) nogil:
+    """K-means combined EM step for one dense data chunk.
+
+    Compute the partial contribution of a single data chunk to the labels and
+    centers.
+    """
+    cdef:
+        int n_samples = labels.shape[0]
+        int n_clusters = centers_old.shape[0]
+        int n_features = centers_old.shape[1]
+
+        floating sq_dist, min_sq_dist
+        int i, j, k, label
+
+    # Instead of computing the full pairwise squared distances matrix,
+    # ||X - C||² = ||X||² - 2 X.C^T + ||C||², we only need to store
+    # the - 2 X.C^T + ||C||² term since the argmin for a given sample only
+    # depends on the centers.
+    # pairwise_distances = ||C||²
+    for i in range(n_samples):
+        for j in range(n_clusters):
+            pairwise_distances[i * n_clusters + j] = centers_squared_norms[j]
+
+    # pairwise_distances += -2 * X.dot(C.T)
+    _gemm(RowMajor, NoTrans, Trans, n_samples, n_clusters, n_features,
+          -2.0, X, n_features, &centers_old[0, 0], n_features,
+          1.0, pairwise_distances, n_clusters)
+
+    for i in range(n_samples):
+        min_sq_dist = pairwise_distances[i * n_clusters]
+        label = 0
+        for j in range(1, n_clusters):
+            sq_dist = pairwise_distances[i * n_clusters + j]
+            if sq_dist < min_sq_dist:
+                min_sq_dist = sq_dist
+                label = j
+        labels[i] = label
+
+        if update_centers:
+            weight_in_clusters[label] += sample_weight[i]
```

Comment:
Should the weight of the previous label be decremented?

Reply:
At the start of an iteration `weight_in_clusters` is full zeros. Then for each sample, we find the closest center, thus the label, and increment the amount of weight for this center.